service/tablet_allocator.cc

```diff
@@ -500,24 +502,36 @@ class load_balancer {
         std::vector<table_id_and_size_desc> tables_need_resize;
         std::vector<table_id_and_size_desc> tables_being_resized;
 
-        static bool table_needs_merge(const table_size_desc& d) {
-            // The initial_tablet_count is respected while the table is in "growing mode".
-            // We say that a table leaves this mode if it required a split above the initial
-            // tablet count. After that, we can rely purely on the average size to say that
-            // a table is shrinking and requires merge.
-            // FIXME: this is not perfect and we may want to leave the mode too if we detect
-            //  average size is decreasing significantly, before any split happened.
-            bool left_growing_mode = !d.resize_decision.initial_decision();
-            lblogger.debug("table_needs_merge: tablet_count={}, avg_tablet_size={}, left_growing_mode={} (seq number: {})",
-                           d.tablet_count, d.avg_tablet_size, left_growing_mode, d.resize_decision.sequence_number);
-            return left_growing_mode && d.tablet_count > 1 && d.avg_tablet_size < d.target_min_tablet_size();
-        }
-        static bool table_needs_split(const table_size_desc& d) {
-            return d.avg_tablet_size > d.target_max_tablet_size();
+        static locator::resize_decision to_resize_decision(const table_size_desc& d) {
+            locator::resize_decision decision;
+
+            auto target_tablet_count = d.tablet_count;
+            // Split based on min_tablet_count or avg_tablet_size, or
+            // if the current resize_decision is split, apply hysteresis,
+            // so it would get cancelled only when crossing back the half-way point.
+            if (d.tablet_count < d.min_tablet_count || d.avg_tablet_size > d.target_max_tablet_size() ||
+                (d.resize_decision.is_split() && d.avg_tablet_size >= d.target_tablet_size)) {
+                // TODO: extend to n-way split when needed
+                target_tablet_count *= 2;
+                decision.way = resize_decision::split{};
+            } else {
+                // Consider merge, as long as it doesn't violate min_tablet_count, or
+                // if the current resize_decision is merge, apply hysteresis,
+                // so it would get cancelled only when crossing back the half-way point.
+                if ((d.avg_tablet_size < d.target_min_tablet_size() && target_tablet_count / 2 >= d.min_tablet_count) ||
+                    (d.resize_decision.is_merge() && d.avg_tablet_size <= d.target_tablet_size)) {
```

Comment:
I think `target_tablet_count / 2 >= d.min_tablet_count` should be a common conjunction term, so that it cancels the hysteresis-induced merge too. Otherwise, post-merge we would split immediately to satisfy min_tablet_count. hints take precedence over size-based count.

Reply:
That's a good point.
If min_tablet_count was changed in the middle of a merge, the new value may indeed cancel it.

Reply:
Fixed in v17 ([021fc3c](https://github.com/scylladb/scylladb/commit/021fc3c75626872957c0c844ade108d68b582c3e))